### PR TITLE
Upgrade eslint + flow

### PR DESCRIPTION
**Pull request by lukyth**
**Summary**

I want to upgrade ESLint and Flow on this project. I ended up with
- Remove unused `eslint-plugin-relay`
- Upgrade all ESLint packages to the latest version and fix all the new lint errors afterward.
  - `eslint-config-fbjs`
  - `eslint-config-prettier`
  - `eslint-plugin-babel`
  - `eslint-plugin-flowtype`
  - `eslint-plugin-jsx-a11y`
  - `eslint-plugin-prettier`
  - `eslint-plugin-react`

- Upgrade Flow (`flow-bin`) to the `0.115.0`.

~Then I found that `yarn test` is failing at~
```
src/component/handlers/composition/__tests__/DraftEditorCompostionHandler-test.js: Support for the experimental syntax 'nullishCoalescingOperator' isn't currently enabled (70:40)
```
~So I added `@babel/plugin-proposal-nullish-coalescing-operator` to jest preprocessor as well.~
Move plugins in `.babelrc` to jest preprocessor as https://github.com/facebook/draft-js/pull/2083#issuecomment-502032839

**Test Plan**

`yarn lint`, `yarn flow`, and `yarn test` should all pass.

### CODE REVIEW

**Comment by claudiopro** [comment]
Hi @lukyth, thanks for contributing a fix to Draft.js! 🎉 Unfortunately I had already started working on this change with #2076 a few days earlier, but I like your change to add the babel plugin to the Jest preprocessor better:

```diff
  process(src, filename) {
    var options = {
      presets: [fbjsConfigurePreset({rewriteModules: {map: moduleMap}})],
+     plugins: [require('@babel/plugin-proposal-nullish-coalescing-operator')],
      filename: filename,
      retainLines: true,
    };
```

whereas I added a `.babelrc` file. Do you mind rebasing and proposing this change again? I'd really love to get this added and give you credit in the commit history.

File: src/model/immutable/ContentBlockNode.js [hunk 1]
```diff
@@ -76,8 +76,7 @@ const decorateCharacterList = (
   return config;
 };
 
-class ContentBlockNode extends (Record(defaultRecord): any)
-  implements BlockNode {
+class ContentBlockNode extends Record(defaultRecord) implements BlockNode {
```
Comment (claudiopro) [inline src/model/immutable/ContentBlockNode.js]: Unfortunately the coercion to `any` is required for internal Facebook Flow typechecking on the www repo, so it has to stay like that for now 😞

Comment (lukyth) [reply src/model/immutable/ContentBlockNode.js]: I see. Got it :)

**Comment by lukyth** [comment]
Hi @claudiopro :) Thank you for your comments! I rebased the branch against `master` and brought back the type coercion. Please have a look again 🙏

File: package.json [hunk 2]
```diff
@@ -32,6 +32,7 @@
     "test-ci": "cross-env NODE_ENV=test npm run lint && npm run flow && npm run test"
   },
   "dependencies": {
+    "@babel/plugin-proposal-nullish-coalescing-operator": "^7.4.4",
```
Comment (claudiopro) [inline package.json]: This is in `devDependencies` already, please remove from `dependencies`. This field is only meant for runtime dependencies of this package.

File: package.json [hunk 3]
```diff
@@ -50,16 +51,15 @@
...
+    "eslint-config-fbjs": "^3.1.0",
+    "eslint-config-prettier": "^4.3.0",
+    "eslint-plugin-babel": "^5.3.0",
+    "eslint-plugin-flowtype": "^3.10.3",
+    "eslint-plugin-jsx-a11y": "^6.2.1",
+    "eslint-plugin-prettier": "^3.1.0",
+    "eslint-plugin-react": "^7.13.0",
     "fbjs-scripts": "^1.1.0",
-    "flow-bin": "^0.92.0",
+    "flow-bin": "^0.98.1",
```
Comment (claudiopro) [inline package.json]: Latest one is `flow-bin@0.101.0`, can you update it? Thanks!

File: src/.flowconfig [hunk 4]
```diff
@@ -37,4 +37,4 @@ suppress_comment=\\(.\\|\n\\)*\\$FlowExpectedError
 module.name_mapper='ReactDOM' -> 'react-dom'
 
 [version]
-^0.92.0
+^0.98.1
```
Comment (claudiopro) [inline src/.flowconfig]: See above.

File: src/model/modifier/__tests__/AtomicBlockUtils-test.js [hunk 5]
```diff
@@ -36,6 +36,7 @@ const getInvariantViolation = msg => {
     /* eslint-disable fb-www/sprintf-like-args */
     invariant(false, msg);
     /* eslint-enable fb-www/sprintf-like-args */
+    return 'Should not reach here.';
```
Comment (claudiopro) [inline src/model/modifier/__tests__/AtomicBlockUtils-test.js]: Is this needed?

File: .eslintrc.js [hunk 6]
```diff
@@ -14,7 +14,7 @@ module.exports = {
     'prettier/standard',
   ],
   rules: {
-    'prettier/prettier': ['error', 'fb'],
```
Comment (claudiopro) [inline .eslintrc.js]: Why is this needed?

File: src/model/modifier/__tests__/AtomicBlockUtils-test.js (hunk 5 above)
Comment (lukyth) [reply src/model/modifier/__tests__/AtomicBlockUtils-test.js]: There'll be an ESLint error without it. `Expected to return a value at the end of arrow function. eslint(consistent-return)`

It expects the function to return something from both `try` and `catch` blocks.
```js
const getInvariantViolation = msg => {
  try {
    /* eslint-disable fb-www/sprintf-like-args */
    invariant(false, msg);
    /* eslint-enable fb-www/sprintf-like-args */
    return 'Should not reach here.';
  } catch (e) {
    return e;
  }
};
```

File: .eslintrc.js (hunk 6 above)
Comment (lukyth) [reply .eslintrc.js]: `eslint-plugin-prettier` dropped support for it. ([source](https://github.com/prettier/eslint-plugin-prettier/commit/478c7e5d2165f3e67e893c9a317b602159eaff9c))
> Drop support for specifying prettier options using the "fb" shorthand
string. Replace "fb" with either config in your .prettierrc or the object:
`{ singleQuote: true, trailingComma: 'all', bracketSpacing: false, jsxBracketSameLine: true, parser: 'flow'}`

File: package.json (hunk 3 above)
Comment (lukyth) [reply package.json]: https://github.com/facebook/fbjs/pull/346 to be released first in order to upgrade to a newer version than `0.98.1`. Otherwise, `yarn flow` will output errors from `node_modules/fbjs`. One of the errors:
Screen Shot 2019-06-15 at 02 50 57

**Comment by lukyth** [comment]
@claudiopro I took care of `@babel/plugin-proposal-nullish-coalescing-operator` 😄, but upgrading `flow-bin` to a version newer than `0.98.1` will requires https://github.com/facebook/fbjs/pull/346 to be released first 😞.

**Comment by mrkev** [comment]
Cool stuff! I want to see what the internal linters say about this.

**Comment by lukyth** [comment]
@mrkev Thanks for checking it out :)

**Comment by mrkev** [comment]
There's a linter warning, but it should be fine. I'll try to get this merged today.
